File: Capítulo_4/Pilhas/Resolvendo problemas usando pilhas/Algoritmo conversor de base/Código em Python/baseConverter.py
# Classe Stack
class Stack:
    def __init__(self):
        self._count = 0
        self._items = {}

    def push(self, element):
        self._items[self._count] = element
        self._count += 1
    
    def size(self):
        return len(self._items)
    
    def is_empty(self):
        return len(self._items) == 0
    
    def pop(self):
        if self.is_empty():
            return None
        self._count -= 1
        result = self._items[self._count]
        del self._items[self._count]
        return result
    
    def peek(self):
        if not (self.is_empty()):
            return self._items[self._count - 1]
        else:
            return None
    
    def clear(self):
        self._count = 0
        self._items = {}

    def to_string(self):
        if self.is_empty():
            return ""
        objString = f"{self._items[0]}"
        for i in range(1, self._count):
            objString = f"{objString}, {self._items[i]}"
        return objString
    
# Algoritmo conversor de base
def base_converter(dec_number, base):
    rem_stack = Stack()
    digits = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    number = dec_number
    base_string = ""
    if (not (base >= 2 and base <= 36)):
        return ""
    while (number > 0):
        rem = number % base
        rem_stack.push(rem)
        number = number // base
    while (not rem_stack.is_empty()):
        base_string += digits[rem_stack.pop()]
    return base_string

print(base_converter(100345, 2)) # 11000011111111001
print(base_converter(100345, 8)) # 303771
print(base_converter(100345, 16)) # 187F9
print(base_converter(100345, 35)) # 2BW0

File: Capítulo_4/Pilhas/Resolvendo problemas usando pilhas/Algoritmo conversor de base/baseConverter.js
// Importação da classe Stack
import { Stack } from "../../Proposta para campos de classe na ECMAScript/privateClass.js"

// Algoritmo conversor de base
function baseConverter(decNumber, base) {
    const remStack = new Stack();
    const digits = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    let number = decNumber;
    let rem;
    let baseString = "";
    if (!(base >= 2 && base <= 36)) {
        return "";
    }
    while (number > 0) {
        rem = Math.floor(number % base);
        remStack.push(rem);
        number = Math.floor(number / base);
    }
    while (!remStack.isEmpty()) {
        baseString += digits[remStack.pop()];
    }
    return baseString;
}

// Podemos usar o algoritmo anterior e exibir o seu resultado no console, assim:
console.log(baseConverter(100345, 2)); // 11000011111111001
console.log(baseConverter(100345, 8)); // 303771
console.log(baseConverter(100345, 16)); // 187F9
console.log(baseConverter(100345, 35)); // 2BW0